src/els_compiler_diagnostics.erl

```diff
@@ -50,36 +61,116 @@ diagnostics(Uri) ->
                || IncludePath <- els_config:get(include_paths)
              ],
-  case compile:file(binary_to_list(Path), Includes ++ ?COMPILER_OPTS) of
+  case compile:file(Path, Includes ++ ?COMPILER_OPTS) of
     {ok, _, WS} ->
-      diagnostics(WS, ?DIAGNOSTIC_WARNING);
+      diagnostics(Path, WS, ?DIAGNOSTIC_WARNING);
     {error, ES, WS} ->
-      diagnostics(WS, ?DIAGNOSTIC_WARNING) ++ diagnostics(ES, ?DIAGNOSTIC_ERROR)
+      diagnostics(Path, WS, ?DIAGNOSTIC_WARNING) ++
+        diagnostics(Path, ES, ?DIAGNOSTIC_ERROR)
   end.
 
+-spec parse(uri()) -> [diagnostic()].
+parse(Uri) ->
+  FileName = binary_to_list(els_uri:path(Uri)),
+  {ok, Epp} = epp:open([ {name, FileName}
+                       , {includes, els_config:get(include_paths)}
+                       ]),
+  Res = [diagnostic(range(Line), Module, Desc, ?DIAGNOSTIC_ERROR)
+         || {error, {Line, Module, Desc}} <- epp:parse_file(Epp)],
+  epp:close(Epp),
+  Res.
 
--spec diagnostics([compiler_msg()], severity()) -> [diagnostic()].
-diagnostics(List, Severity) ->
-  lists:flatten([[ diagnostic(location(Line), Module, Desc, Severity)
+%% @doc Convert compiler messages into diagnostics
+%%
+%% Convert a list of compiler messages of a given severity (warning,
+%% error) into a list of diagnostic data structures, as expected by
+%% the LSP protocol.
+%% Compiler messages related to included files are grouped together
+%% and they are presented to the user by highlighting the line where
+%% the file inclusion happens.
+-spec diagnostics(list(), [compiler_msg()], severity()) -> [diagnostic()].
+diagnostics(Path, List, Severity) ->
+  Uri = els_uri:uri(list_to_binary(Path)),
+  {ok, [Document]} = els_dt_document:lookup(Uri),
+  lists:flatten([[ diagnostic( Path
+                             , MessagePath
+                             , range(Line)
+                             , Document
+                             , Module
+                             , Desc
+                             , Severity)
                    || {Line, Module, Desc} <- Info]
-                 || {_Filename, Info } <- List]).
+                 || {MessagePath, Info} <- List]).
 
--spec diagnostic(erl_anno:location(), module(), string(), integer()) ->
+-spec diagnostic( string()
+                , string()
+                , poi_range()
+                , els_dt_document:item()
+                , module()
+                , string()
+                , integer()) -> diagnostic().
+diagnostic(Path, Path, Range, _Document, Module, Desc, Severity) ->
+  %% The compiler message is related to the same .erl file, so
+  %% preserve the location information.
+  diagnostic(Range, Module, Desc, Severity);
+diagnostic(_Path, MessagePath, Range, Document, Module, Desc0, Severity) ->
+  #{from := {Line, _}} = Range,
+  InclusionRange = inclusion_range(MessagePath, Document),
+  %% The compiler message is related to an included file. Replace the
+  %% original location with the location of the file inclusion.
```

Comment:
It is possible to return diagnostics for other files, which are then shown at the appropriate locations in the other file when opened.

So a general strategy should be to show a diagnostic in the .erl file at the location of the include, saying there are problems with the file, but then report diagnostics for the .hrl itself.

I have not run this code, so maybe that does in fact happen.

Reply:
Yes, this is exactly what it does.